diff -ruN openssl-1.0.1f.orig/ssl/s3_lib.c openssl-1.0.1f/ssl/s3_lib.c
--- openssl-1.0.1f.orig/ssl/s3_lib.c	2014-01-06 14:47:42.000000000 +0100
+++ openssl-1.0.1f/ssl/s3_lib.c	2014-01-09 07:56:58.000000000 +0100
@@ -203,6 +203,7 @@
      },
 
 /* Cipher 03 */
+/* disabling 
     {
      1,
      SSL3_TXT_RSA_RC4_40_MD5,
@@ -217,6 +218,7 @@
      40,
      128,
      },
+*/
 
 /* Cipher 04 */
     {
@@ -251,6 +253,7 @@
      },
 
 /* Cipher 06 */
+/* disabling 
     {
      1,
      SSL3_TXT_RSA_RC2_40_MD5,
@@ -265,6 +268,7 @@
      40,
      128,
      },
+*/
 
 /* Cipher 07 */
 #ifndef OPENSSL_NO_IDEA
@@ -285,6 +289,7 @@
 #endif
 
 /* Cipher 08 */
+/* disabling 
     {
      1,
      SSL3_TXT_RSA_DES_40_CBC_SHA,
@@ -299,8 +304,10 @@
      40,
      56,
      },
+*/
 
 /* Cipher 09 */
+/* disabling 
     {
      1,
      SSL3_TXT_RSA_DES_64_CBC_SHA,
@@ -315,6 +322,7 @@
      56,
      56,
      },
+*/
 
 /* Cipher 0A */
     {
@@ -431,6 +439,7 @@
 
 /* The Ephemeral DH ciphers */
 /* Cipher 11 */
+/* disabling 
     {
      1,
      SSL3_TXT_EDH_DSS_DES_40_CBC_SHA,
@@ -445,8 +454,10 @@
      40,
      56,
      },
+*/
 
 /* Cipher 12 */
+/* disabling 
     {
      1,
      SSL3_TXT_EDH_DSS_DES_64_CBC_SHA,
@@ -461,6 +472,7 @@
      56,
      56,
      },
+*/
 
 /* Cipher 13 */
     {
@@ -479,6 +491,7 @@
      },
 
 /* Cipher 14 */
+/* disabling
     {
      1,
      SSL3_TXT_EDH_RSA_DES_40_CBC_SHA,
@@ -493,8 +506,9 @@
      40,
      56,
      },
-
+*/
 /* Cipher 15 */
+/* disabling 
     {
      1,
      SSL3_TXT_EDH_RSA_DES_64_CBC_SHA,
@@ -509,6 +523,7 @@
      56,
      56,
      },
+*/
 
 /* Cipher 16 */
     {
