diff --git a/ssl/s3_clnt.c b/ssl/s3_clnt.c
index bf1ef47..ef638c4 100644
--- a/ssl/s3_clnt.c
+++ b/ssl/s3_clnt.c
@@ -3481,6 +3481,12 @@ int ssl3_check_cert_and_algorithm(SSL *s)
                SSL_R_MISSING_DH_RSA_CERT);
         goto f_err;
     }
+	else if ((alg_k & (SSL_kEDH|SSL_kDHr|SSL_kDHd)) &&
+	         (dh == NULL || DH_size(dh)*8 < 1024))
+		{
+		SSLerr(SSL_F_SSL3_CHECK_CERT_AND_ALGORITHM,SSL_R_WEAK_DH_GROUP);
+		goto f_err;
+		}
 # ifndef OPENSSL_NO_DSA
     else if ((alg_k & SSL_kDHd) && !has_bits(i, EVP_PK_DH | EVP_PKS_DSA)) {
         SSLerr(SSL_F_SSL3_CHECK_CERT_AND_ALGORITHM,
diff --git a/ssl/ssl.h b/ssl/ssl.h
index 472e070..3ea316c 100644
--- a/ssl/ssl.h
+++ b/ssl/ssl.h
@@ -3062,6 +3062,7 @@ void ERR_load_SSL_strings(void);
 # define SSL_R_UNSUPPORTED_SSL_VERSION                    259
 # define SSL_R_UNSUPPORTED_STATUS_TYPE                    329
 # define SSL_R_USE_SRTP_NOT_NEGOTIATED                    369
+# define SSL_R_WEAK_DH_GROUP                              371
 # define SSL_R_WRITE_BIO_NOT_SET                          260
 # define SSL_R_WRONG_CIPHER_RETURNED                      261
 # define SSL_R_WRONG_MESSAGE_TYPE                         262
diff --git a/ssl/ssl_err.c b/ssl/ssl_err.c
index 9889a27..a97b4b5 100644
--- a/ssl/ssl_err.c
+++ b/ssl/ssl_err.c
@@ -612,6 +612,7 @@ static ERR_STRING_DATA SSL_str_reasons[]=
     {ERR_REASON(SSL_R_UNSUPPORTED_SSL_VERSION), "unsupported ssl version"},
     {ERR_REASON(SSL_R_UNSUPPORTED_STATUS_TYPE), "unsupported status type"},
     {ERR_REASON(SSL_R_USE_SRTP_NOT_NEGOTIATED), "use srtp not negotiated"},
+    {ERR_REASON(SSL_R_WEAK_DH_GROUP), "weak dh group"},
     {ERR_REASON(SSL_R_WRITE_BIO_NOT_SET), "write bio not set"},
     {ERR_REASON(SSL_R_WRONG_CIPHER_RETURNED), "wrong cipher returned"},
     {ERR_REASON(SSL_R_WRONG_MESSAGE_TYPE), "wrong message type"},
